Add installer shell script

mavam opened the pull request:
This PR adds an installer shell script so that installing Tenzir will be as simple as:

```bash
curl -L get.tenzir.app | sh
```

This supersedes #2476, a more complex attempt that was Homebrew-inspired.

dit7ya commented on scripts/install.sh:
```diff
@@ -0,0 +1,139 @@
+then
+  echo "Try Tenzir with our ${bold}flake.nix${normal}:"
+  echo
+  echo "    ${bold}nix run github:tenzir/tenzir/stable${normal}"
```
The https://github.com/tenzir/tenzir/tree/stable branch is 500+ commits behind, what is our plan with that @tobim ?

Also, is our cachix working? I tried 

```nix run github:tenzir/tenzir/stable```

but it is building stuff from source. (I have vast cachix enbled).

mavam replied:
It sounds like we want a story for this, to be addressed post alpha testing. As long as the instructions are sound, we can focus on the implementation of this separately.

dit7ya replied:
Cache not working needs a story, yes. But the correct release binary should be part of alpha.

dominiklohmann replied:
I would assume that the cache is broken because of the repository-renaming.

We usually update the `stable` branch when we cut a release. So at least for the Alpha, we would need a different mechanism here.

tobim replied:
> Also, is our cachix working? I tried
>
> nix run github:tenzir/tenzir/stable
>
> but it is building stuff from source. (I have vast cachix enbled).

We can't use cachix for closed source code, since we are still on the free version. (poke @mavam)

mavam replied:
I'm resolving this thread, as the Cachix discussion takes place elsewhere.

rdettai commented:
Following up on https://github.com/tenzir/issues/issues/472#issuecomment-1597476595:

What I would run in CI:
- a matrix with `ubuntu-22.04`, `ubuntu-20.04`, `macos-12`, `macos-11` (if all these are meant to be supported)
- in the job:
  - run the exact command you document, e.g `curl -L get.tenzir.app | sh`
  - run the commands that are documented in the script output, i.e:
```
echo "You're all set! Next steps:"
echo "  - Run a pipeline via ${blue}tenzir <pipeline>${normal}"
echo "  - Spawn a node via ${blue}tenzir-node${normal}" 
```

mavam commented:
> a matrix with ubuntu-22.04, ubuntu-20.04, macos-12, macos-11 (if all these are meant to be supported)

It would obviously be nice to cover all platforms. But for macOS, we don't have packages yet. Only for Linux. Can we get a Debian and non-Debian from GitHub?

> run the exact command you document, e.g `curl -L get.tenzir.app | sh`

@rdettai this won't work because the installer lives in the repo. You will never get that version that lives in the PR. That said, I don't think it's important that we go through the DNS lookup and CNAME redirect. We can simply invoke `install.sh` from CI.

rdettai commented:
> Can we get a Debian and non-Debian from GitHub
I think the way to do this would be through a matrix of docker images: https://github.com/orgs/community/discussions/25079#discussioncomment-3246416

>  this won't work because the installer lives in the repo. You will never get that version that lives in the PR. That said, I don't think it's important that we go through the DNS lookup and CNAME redirect. We can simply invoke install.sh from CI

We could also decouple the two:
- have a test ensuring that the DNS setup works by just ensuring we get a 200 with something that looks like a script in it
- run install.sh

That being said, if the DNS setup is fully exogenous from the code base, it should also be probed separately. It would rather go into a sort of health bot that goes through all Tenzir resources (docs, website...) and checks their health (no dead links...). So I think that it would be enough for this PR to validate that the script works.

dominiklohmann commented:
> Does this go into the `vast.yaml` workflow? Or in a separate file? (cc @dominiklohmann)

A separate workflow because it has separate triggers.

In any case, what's the urgency of having this tested in CI? Can we add that at a later point in time? I think it's not quite so simple before we have a new release.

mavam commented:
Agreed, lets' factor the CI discussion out of this PR. Get this merged in a working version, and then tackle the topic later.

mavam commented:
Requesting review from @dominiklohmann for general onboarding flow and @tobim for ops details.

tobim commented on scripts/install.sh:
```diff
@@ -0,0 +1,178 @@
+if [ "${PLATFORM}" = "Debian" ]
+then
+  action "Installing via dpkg"
+  dpkg -i "${PACKAGE}"
```
If I run this as a regular user it fails, but continues with the script, which then fails more. We should `set -e` in the script and we should `sudo` here.

tobim replied:
```suggestion
  sudo dpkg -i "${tmpdir}/${PACKAGE}"
```

tobim commented on scripts/install.sh:
```diff
@@ -0,0 +1,178 @@
+echo "available. Visit ${bold}https://tenzir.com/pricing${normal} for a" \
+  "feature comparison.)"
+echo
+echo "Press ${green}ENTER${normal} to proceed with the installation."
```
We need to tell the user what is going to happen next here, depending on the detected platform.

Something like
```suggestion
echo "I will now run \`sudo dpkg -i tenzir-linux-static.deb\` to finish the installation."
echo "Press ${green}ENTER${normal} to proceed."
```

tobim reviewed (changes requested):
As a stylistic suggestion: Script variables that are not exported as environment variables should be all lowercase instead of ALLCAPS.

tobim commented on scripts/install.sh:
```diff
@@ -0,0 +1,178 @@
+fi
+echo "Using ${PACKAGE}"
+
+# Download package.
+BASE="https://github.com/tenzir/tenzir/releases/latest/download"
+URL="${BASE}/${PACKAGE}"
+action "Downloading ${URL}"
+curl --progress-bar -L "${URL}" -o "${PACKAGE}" || exit 1
```
`curl` is not installed by default in Debian, but `wget` is. I suspect that is the case in some other distros as well.
```suggestion

check() {
  command -v "$1" 2>&1 >/dev/null
}

# Download package.
tmpdir="$(dirname "$(mktemp -u --tmpdir)")"
BASE="https://github.com/tenzir/tenzir/releases/latest/download"
URL="${BASE}/${PACKAGE}"
action "Downloading ${URL}"
if check quizf; then
  exit 1
elif check wget; then
  wget -q --show-progress -O "${tmpdir}/${PACKAGE}" "${URL}"
elif check curl; then
  curl --progress-bar -L -o "${tmpdir}/${PACKAGE}" "${URL}"
else
  echo "Neither 'wget' nor 'curl' in \$PATH, aborting..."
  exit 1
fi
```

mavam replied:
`quizf` doesn't exist on macOS. (Never heard of it before.) I think `curl` and `wget` are enough, no?

mavam replied:
I've switched to `wget` and `curl`. Feel free to resolve if that suits the bill.

tobim replied:
That was a copy-n-paste mistake. I just wanted to make sure the code worked as intended before posting the suggestion.

tobim commented on scripts/install.sh:
```diff
@@ -0,0 +1,178 @@
+  action "Installing via dpkg"
+  dpkg -i "${PACKAGE}"
+  action "Checking node status"
+  systemctl status tenzir || exit 1
```
```suggestion
  sudo systemctl status tenzir || exit 1
```

tobim commented on scripts/install.sh:
```diff
@@ -0,0 +1,178 @@
+elif [ "${PLATFORM}" = "Linux" ]
+then
+  action "Unpacking tarball"
+  tar xzf "${PACKAGE}" -C /
```
```suggestion
  sudo tar xzf "${tmpdir}/${PACKAGE}" -C /
```

mavam replied:
Should we trap EXIT and delete `tmpdir`?

mavam commented:
@tobim I've implemented your requests. A few things are noteworthy:

1. I simplified `mktemp` because `--tempdir` is not a thing on macOS
2. I don't think we need to announce what will happen, but rather be transparent what happens. Otherwise we'd need user input on everything.
3. I ignored `quizf` because I've never heard of it.

dominiklohmann commented on scripts/install.sh:
```diff
@@ -0,0 +1,197 @@
+  exit 1
+fi
+action "Installing package into ${prefix}"
+if [ "${platform}" = "Debian" ]
+then
+  action "Installing via dpkg"
+  sudo dpkg -i "${tmpdir}/${package}"
+  action "Checking node status"
+  sudo systemctl status tenzir
+elif [ "${platform}" = "Linux" ]
+then
+  action "Unpacking tarball"
+  sudo tar xzf "${tmpdir}/${package}" -C /
+fi
```
These two behave quite differently. The former should tell the user that a systemd service was set up for them, and what to do next.

mavam replied:
Yes, that's unavoidable. We have a guidance section at the end where we should provide specific instructions on what to do next. The text in the middle is not a good place to provide instructions.

I'm already mentioning there that you should start a node if you've used a tarball. I don't think we should mention systemd if it might not be there. We can assume that users can take it from here IMO.

dominiklohmann replied:
The difference is that one starts a node, and the other does not. Can we unify that behavior somehow?

mavam replied:
I don't think so. It's expected to start a service after `dpkg -i`, but we can't just start a node with service manager that we don't know for the generic case. Instructions are the only way to go here.

mavam replied:
We'll improve the messaging over time. Let's start with what we have an wait for feedback.

dominiklohmann commented on scripts/install.sh:
```diff
@@ -0,0 +1,197 @@
+then
+  platform=macOS
+fi
+echo "Found ${platform}."
```
If platform is empty at this point then you should exit with a helpful error message.

mavam replied:
`uname -s` doesn't return anything empty on anything reasonable we support.

mavam replied:
We're dispatching on it in the next block and bail out if there's no appropriate package, so this is handled correctly.

dominiklohmann replied:
To program this a bit more defensively, could you add an `else` that prints a logic error and exits or an assertion here? That will make that immediately visible to the reader of this code.

mavam replied:
Added that now.

dominiklohmann reviewed (commented):
I mostly looked at the flow and had some minor things that stood out to me. I'll leave the final approval to @tobim.

tobim reviewed (approved):
> I don't think we need to announce what will happen, but rather be transparent what happens. Otherwise we'd need user input on everything.

My suggestion does not add more steps, just a little bit of transparency so the user has more information to decide whether they want to `<enter>` or `<ctrl-c>`.

mavam commented on scripts/install.sh:
```diff
@@ -0,0 +1,197 @@
+# Download package.
+base="https://github.com/tenzir/tenzir/releases/latest/download"
+url="${base}/${package}"
+tmpdir="$(mktemp -u)"
```
Makes sense, fixed that.

dominiklohmann commented on scripts/install.sh:
```diff
@@ -0,0 +1,197 @@
+set -u
+
+# Abort on error.
+set -e
```
If you can use bash rather than sh, it would be helpful to have `set -euo pipefail` here. That usually helps a lot with checking correctness.

mavam replied:
We're not using Bash on purpose.

dominiklohmann replied:
Why is that? `/usr/bin/env bash` is available basically everywhere.

mavam replied:
There's no need in using Bash just for `pipefail`. POSIX Shell makes the script far more accessible.

mavam replied:
Resolving this for now. We can reconsider Bash if we need more advanced scripting fu.
